wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala

```diff
@@ -0,0 +1,268 @@
+    }
+
+  }
+
+  private[wallet] case class ProcessTxResult(
+      updatedIncoming: List[SpendingInfoDb],
+      updatedOutgoing: List[SpendingInfoDb])
+
+  /////////////////////
+  // Internal wallet API
+
+  /**
+    * Processes TXs originating from our wallet.
+    * This is called right after we've signed a TX,
+    * updating our UTXO state.
+    */
+  private[wallet] def processOurTransaction(
+      transaction: Transaction,
+      confirmations: Int): Future[ProcessTxResult] = {
+    logger.info(
+      s"Processing TX from our wallet, transaction=${transaction.txIdBE} with confirmations=$confirmations")
+    processTransactionImpl(transaction, confirmations).map { result =>
+      val txid = transaction.txIdBE
+      val changeOutputs = result.updatedIncoming.length
+      val spentOutputs = result.updatedOutgoing.length
+
+      logger.info(
+        s"Processing of internal transaction=$txid resulted in changeOutputs=$changeOutputs and spentUTXOs=$spentOutputs")
+      result
+    }
+  }
+
+  /////////////////////
+  // Private methods
+
+  /** Does the grunt work of processing a TX.
+    * This is called by either the internal or public TX
+    * processing method, which logs and transforms the
+    * output fittingly.
+    */
+  private def processTransactionImpl(
+      transaction: Transaction,
+      confirmations: Int): Future[ProcessTxResult] = {
+
+    val incomingTxoFut: Future[Vector[SpendingInfoDb]] =
+      spendingInfoDAO
+        .findTx(transaction)
+        .flatMap {
+          // no existing elements found
+          case Vector() =>
+            processNewIncomingTx(transaction, confirmations).map(_.toVector)
+
+          case txos: Vector[SpendingInfoDb] =>
+            val txoProcessingFutures =
+              txos
+                .map(processExistingIncomingTxo(transaction, confirmations, _))
+
+            Future
+              .sequence(txoProcessingFutures)
+
+        }
+
+    val outgoingTxFut: Future[Vector[SpendingInfoDb]] = {
+      for {
+        outputsBeingSpent <- spendingInfoDAO.findOutputsBeingSpent(transaction)
+        processed <- FutureUtil.sequentially(outputsBeingSpent)(
+          markAsSpentIfUnspent)
+      } yield processed.flatten.toVector
+
+    }
+
+    val aggregateFut =
+      for {
+        incoming <- incomingTxoFut
+        outgoing <- outgoingTxFut
+      } yield {
+        ProcessTxResult(incoming.toList, outgoing.toList)
+      }
+
+    aggregateFut.failed.foreach { err =>
+      val msg = s"Error when processing transaction=${transaction.txIdBE}"
+      logger.error(msg, err)
+    }
+
+    aggregateFut
+  }
+
+  /** If the given UTXO is marked as unspent, updates
+    * its spending status. Otherwise returns `None`.
+    */
+  private val markAsSpentIfUnspent: SpendingInfoDb => Future[
+    Option[SpendingInfoDb]] = { out =>
+    if (out.spent) {
+      Future.successful(None)
+    } else {
+      val updatedF =
+        spendingInfoDAO.update(out.copyWithSpent(spent = true))
```

Comment:
So I actually don't think we want to do the sql update here, we probably want to do it in the caller of this method, so that the transaction will be updated atomically. In the current implementation, we could have some `SpendingInfoDb` be marked as spent, while some others not marked as spent within the same transaction because this database call failed for whatever reason.

TLDR, we probably want to make this update using `.transactionally`.

Reply:
Made a change where `markAsSpentIfUnspent` now takes in a sequence of TXOs, and calls `updateAll` in its DAO. `CRUD.updateAll` is also updated to process elements transactionally. 

Is this a solution you think works?